File: bench/split_join/split_join.java
// Java peer of bench/split_join.vader. `String.split` takes a regex, but the
// single-character fast path skips the engine — which is what Java code writes
// for a separator this simple.

import java.util.ArrayList;
import java.util.List;

static final int FIELD_COUNT = 20_000;
static final int PASS_COUNT = 24;
static final String SEPARATOR = ",";
static final long STRIDE = 1103515245L;

static String buildText() {
    List<String> parts = new ArrayList<>(FIELD_COUNT);
    for (int index = 0; index < FIELD_COUNT; index++) {
        parts.add(Long.toString((long) index * STRIDE % 1000003L));
    }
    return String.join(SEPARATOR, parts);
}

void main() {
    String text = buildText();
    long pieces = 0, joinedLength = 0;
    for (int pass = 0; pass < PASS_COUNT; pass++) {
        String[] parts = text.split(SEPARATOR);
        List<String> kept = new ArrayList<>(parts.length);
        for (String part : parts) {
            if (part.length() % 2 == 1) kept.add(part);
        }
        pieces += kept.size();
        joinedLength += String.join(SEPARATOR, kept).length();
    }
    IO.println("split_join fields=%d passes=%d pieces=%d length=%d"
        .formatted(FIELD_COUNT, PASS_COUNT, pieces, joinedLength));
}

File: bench/split_join/split_join.ts
// TypeScript peer of bench/split_join.vader. `String.prototype.split` and
// `Array.prototype.join` are the idiomatic pair.

const FIELD_COUNT = 20_000;
const PASS_COUNT = 24;
const SEPARATOR = ",";
const STRIDE = 1103515245;

function buildText(): string {
  const parts: string[] = [];
  for (let index = 0; index < FIELD_COUNT; index++) {
    parts.push(String((index * STRIDE) % 1000003));
  }
  return parts.join(SEPARATOR);
}

const text = buildText();
let pieces = 0;
let joinedLength = 0;
for (let pass = 0; pass < PASS_COUNT; pass++) {
  const parts = text.split(SEPARATOR);
  const kept: string[] = [];
  for (let i = 0; i < parts.length; i++) {
    if (parts[i]!.length % 2 === 1) kept.push(parts[i]!);
  }
  pieces += kept.length;
  joinedLength += kept.join(SEPARATOR).length;
}
console.log(
  `split_join fields=${FIELD_COUNT} passes=${PASS_COUNT} pieces=${pieces} length=${joinedLength}`,
);
